api/ombpdf/semdb.py

```diff
@@ -1,12 +1,153 @@
-from lxml import etree
+import logging
 
-from document.xml_importer.importer import import_xml_doc
+from document.models import DocNode, FootnoteCitation
+from document.tree import DocCursor
 
-from . import semdom
+from . import semtree
+
+logger = logging.getLogger(__name__)
+
+
+class PDFAwareCursor(DocCursor):
+    """Extension of DocCursor which also tracks the PDF element which created
+    each node."""
+    @property
+    def pdf_node(self):
+        return self.tree.node[self.identifier].get('pdf_node')
+
+    @pdf_node.setter
+    def pdf_node(self, value):
+        self.tree.node[self.identifier]['pdf_node'] = value
 
 
 def to_db(doc, policy):
-    dom = semdom.to_dom(doc)
-    xml = etree.fromstring(dom.toxml())
-    root = import_xml_doc(policy, xml, ignore_preamble=True)
-    return root
+    writer = DatabaseWriter(policy)
+    builder = semtree.SemanticTreeBuilder(doc, writer)
+    builder.build()
+
+    writer.cursor.nested_set_renumber()
+    # Account for a Django bug which prevents us from saving these
+    # directly due to saving the DocNodes in a bulk_insert
```

Comment:
Is this Django bug recorded in their issue tracker or anything? I thought it might be useful to link to that, both for additional context, but also so that someday we might follow it to find it fixed, and thus be able to remove this workaround.

Reply:
I'll try to track that down and add a comment.